File: VB/WindowsApplication3/TypeConverterHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Windows.Forms
Imports DevExpress.Skins
Imports System.Drawing
Imports System.ComponentModel
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Columns
Imports System.Reflection
Imports System.Collections.ObjectModel
Imports DevExpress.XtraEditors.Repository
Imports System.Collections

Namespace DXSample
	Public Class TypeConverterHelper
		Private edit As RepositoryItemGridLookUpEdit
		Private unboundColumnFieldName As String = "UnboundColumn", convertedProperty As String = String.Empty

		Public Sub New(ByVal edit As RepositoryItemGridLookUpEdit, ByVal unboundColumnFieldName As String, ByVal convertedProperty As String)
			Me.edit = edit
			Me.unboundColumnFieldName = unboundColumnFieldName
			Me.convertedProperty = convertedProperty

			CreateUnboundColumn(edit.View, unboundColumnFieldName)

		End Sub

		Private Sub SubscribeToEvents()
			AddHandler edit.View.CustomUnboundColumnData, AddressOf OnCustomUnboundColumnData
		End Sub

		Private Sub CreateUnboundColumn(ByVal view As GridView, ByVal unboundColumnFieldName As String)
			edit.View.PopulateColumns()
			Dim originColumn As GridColumn = view.Columns(convertedProperty)
			If originColumn IsNot Nothing Then
				originColumn.Visible = False
				Dim column As GridColumn = view.Columns.AddField(unboundColumnFieldName)
				column.UnboundType = DevExpress.Data.UnboundColumnType.Object
				column.VisibleIndex = view.VisibleColumns.Count
				SubscribeToEvents()
			End If
		End Sub

		Private Sub OnCustomUnboundColumnData(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.CustomColumnDataEventArgs)
			 If e.Column.FieldName = unboundColumnFieldName Then
				 If e.IsGetData Then
					 Dim dataSource As IList = TryCast(edit.DataSource, IList)
					 Dim obj As Object = dataSource(e.ListSourceRowIndex)
					 If obj Is Nothing Then
						 Return
					 End If
					 Dim descriptor As PropertyDescriptor = TypeDescriptor.GetProperties(obj.GetType())(convertedProperty)
					 If descriptor Is Nothing Then
						 Return
					 End If
					 Dim value As Object = descriptor.GetValue(obj)
					 Dim converter As TypeConverter = descriptor.Converter
					 If converter IsNot Nothing AndAlso converter.CanConvertFrom(value.GetType()) Then
						 e.Value = converter.ConvertFrom(value)
					 End If
				 End If
			 End If
		End Sub

		Private Sub UnsubcribeFromEvents()
			RemoveHandler edit.View.CustomUnboundColumnData, AddressOf OnCustomUnboundColumnData
		End Sub

		Public Sub RemoveUnboundColumn()
			Dim column As GridColumn = edit.View.Columns(unboundColumnFieldName)
			Dim originColumn As GridColumn = edit.View.Columns(convertedProperty)
			If column IsNot Nothing AndAlso originColumn IsNot Nothing Then
				edit.View.Columns.Remove(column)
				originColumn.Visible = True
				UnsubcribeFromEvents()
			End If
		End Sub
	End Class
End Namespace

File: CS/WindowsApplication3/TypeConverterHelper.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using DevExpress.Skins;
using System.Drawing;
using System.ComponentModel;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Columns;
using System.Reflection;
using System.Collections.ObjectModel;
using DevExpress.XtraEditors.Repository;
using System.Collections;

namespace DXSample {
    public class TypeConverterHelper {
        RepositoryItemGridLookUpEdit edit;
        string unboundColumnFieldName = "UnboundColumn",
            convertedProperty = string.Empty;

        public TypeConverterHelper(RepositoryItemGridLookUpEdit edit, string unboundColumnFieldName,
            string convertedProperty){
            this.edit = edit;
            this.unboundColumnFieldName = unboundColumnFieldName;
            this.convertedProperty = convertedProperty;

            CreateUnboundColumn(edit.View, unboundColumnFieldName);
           
        }

        void SubscribeToEvents() {
            edit.View.CustomUnboundColumnData += OnCustomUnboundColumnData;
        }

        private void CreateUnboundColumn(GridView view, string unboundColumnFieldName) {
            edit.View.PopulateColumns();
            GridColumn originColumn = view.Columns[convertedProperty];
            if (originColumn != null)
            {
                originColumn.Visible = false;
                GridColumn column = view.Columns.AddField(unboundColumnFieldName);
                column.UnboundType = DevExpress.Data.UnboundColumnType.Object;
                column.VisibleIndex = view.VisibleColumns.Count;
                SubscribeToEvents();
            }
        }

        void OnCustomUnboundColumnData(object sender, DevExpress.XtraGrid.Views.Base.CustomColumnDataEventArgs e) {
             if(e.Column.FieldName == unboundColumnFieldName) {
                 if(e.IsGetData) {
                     IList dataSource = edit.DataSource as IList;
                     object obj = dataSource[e.ListSourceRowIndex];
                     if (obj == null) return;
                     PropertyDescriptor descriptor = TypeDescriptor.GetProperties(obj.GetType())[convertedProperty];
                     if(descriptor == null) return;
                     object value = descriptor.GetValue(obj);
                     TypeConverter converter = descriptor.Converter;
                     if(converter != null && converter.CanConvertFrom(value.GetType()))
                         e.Value = converter.ConvertFrom(value);
                 }
             }
        }

        void UnsubcribeFromEvents() {
            edit.View.CustomUnboundColumnData -= OnCustomUnboundColumnData;
        }

        public void RemoveUnboundColumn()
        {
            GridColumn column = edit.View.Columns[unboundColumnFieldName];
            GridColumn originColumn = edit.View.Columns[convertedProperty];
            if (column != null && originColumn != null)
            {
                edit.View.Columns.Remove(column);
                originColumn.Visible = true;
                UnsubcribeFromEvents();
            }
        }
    }
}
    